File: snippets/screens/flutter/following_screen/2/following_screen.dart
import 'package:flutter/material.dart';

class FollowingScreen extends StatefulWidget {
  final String username;

  const FollowingScreen({Key? key, required this.username}) : super(key: key);

  @override
  _FollowingScreenState createState() => _FollowingScreenState();
}

class _FollowingScreenState extends State<FollowingScreen> {
  final TextEditingController _searchController = TextEditingController();
  List<User> following = [
    User(
      id: '1',
      username: 'alice_wonder',
      displayName: 'Alice Wonder',
      avatar: 'https://example.com/avatar3.jpg',
      isFollowing: true,
      isFollowedBy: true,
      followedDate: '2 months ago',
    ),
    User(
      id: '2',
      username: 'bob_builder',
      displayName: 'Bob Builder',
      avatar: 'https://example.com/avatar4.jpg',
      isFollowing: true,
      isFollowedBy: false,
      followedDate: '1 month ago',
    ),
    User(
      id: '3',
      username: 'charlie_brown',
      displayName: 'Charlie Brown',
      avatar: 'https://example.com/avatar5.jpg',
      isFollowing: true,
      isFollowedBy: true,
      followedDate: '3 weeks ago',
    ),
    User(
      id: '4',
      username: 'diana_prince',
      displayName: 'Diana Prince',
      avatar: 'https://example.com/avatar6.jpg',
      isFollowing: true,
      isFollowedBy: false,
      followedDate: '2 weeks ago',
    ),
    User(
      id: '5',
      username: 'edward_stark',
      displayName: 'Edward Stark',
      avatar: 'https://example.com/avatar7.jpg',
      isFollowing: true,
      isFollowedBy: true,
      followedDate: '1 week ago',
    ),
    User(
      id: '6',
      username: 'frank_castle',
      displayName: 'Frank Castle',
      avatar: 'https://example.com/avatar8.jpg',
      isFollowing: true,
      isFollowedBy: false,
      followedDate: '3 days ago',
    ),
  ];

  List<User> filteredFollowing = [];
  String selectedCategory = 'All';
  bool isLoading = false;

  @override
  void initState() {
    super.initState();
    filteredFollowing = List.from(following);
    _searchController.addListener(_filterFollowing);
  }

  void _filterFollowing() {
    final query = _searchController.text.toLowerCase();
    setState(() {
      List<User> baseList = _getFilteredByCategory();

      if (query.isEmpty) {
        filteredFollowing = baseList;
      } else {
        filteredFollowing = baseList.where((user) {
          return user.username.toLowerCase().contains(query) ||
                 user.displayName.toLowerCase().contains(query);
        }).toList();
      }
    });
  }

  List<User> _getFilteredByCategory() {
    switch (selectedCategory) {
      case 'Following you back':
        return following.where((user) => user.isFollowedBy).toList();
      case 'Not following back':
        return following.where((user) => !user.isFollowedBy).toList();
      case 'Recently followed':
        // Sort by most recent (in real app, use actual dates)
        List<User> recent = List.from(following);
        recent.sort((a, b) => a.followedDate.compareTo(b.followedDate));
        return recent.take(3).toList();
      default:
        return following;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text('Following'),
            Text(
              '${following.length} following',
              style: TextStyle(fontSize: 14, color: Colors.grey[600]),
            ),
          ],
        ),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        actions: [
          PopupMenuButton<String>(
            icon: Icon(Icons.sort),
            onSelected: _sortFollowing,
            itemBuilder: (context) => [
              PopupMenuItem(value: 'alphabetical', child: Text('Sort A-Z')),
              PopupMenuItem(value: 'recent', child: Text('Recently followed')),
              PopupMenuItem(value: 'earliest', child: Text('Followed earliest')),
            ],
          ),
        ],
      ),
      body: Column(
        children: [
          // Search bar
          Padding(
            padding: EdgeInsets.all(16),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                hintText: 'Search following...',
                prefixIcon: Icon(Icons.search),
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(8),
                  borderSide: BorderSide(color: Colors.grey[300]!),
                ),
                filled: true,
                fillColor: Colors.grey[100],
              ),
            ),
          ),
          // Categories
          Container(
            height: 50,
            child: ListView(
              scrollDirection: Axis.horizontal,
              padding: EdgeInsets.symmetric(horizontal: 16),
              children: [
                _buildCategoryChip('All'),
                _buildCategoryChip('Following you back'),
                _buildCategoryChip('Not following back'),
                _buildCategoryChip('Recently followed'),
              ],
            ),
          ),
          SizedBox(height: 8),
          // Following list
          Expanded(
            child: RefreshIndicator(
              onRefresh: _refreshFollowing,
              child: ListView.builder(
                itemCount: filteredFollowing.length,
                itemBuilder: (context, index) {
                  return _buildFollowingItem(filteredFollowing[index]);
                },
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildCategoryChip(String label) {
    bool isSelected = selectedCategory == label;
    return Padding(
      padding: EdgeInsets.only(right: 8),
      child: FilterChip(
        label: Text(
          label,
          style: TextStyle(
            color: isSelected ? Theme.of(context).primaryColor : Colors.grey[700],
            fontWeight: isSelected ? FontWeight.bold : FontWeight.normal,
          ),
        ),
        selected: isSelected,
        onSelected: (selected) {
          setState(() {
            selectedCategory = label;
          });
          _filterFollowing();
        },
        selectedColor: Theme.of(context).primaryColor.withOpacity(0.1),
        backgroundColor: Colors.grey[100],
        side: BorderSide(
          color: isSelected ? Theme.of(context).primaryColor : Colors.grey[300]!,
        ),
      ),
    );
  }

  Widget _buildFollowingItem(User user) {
    return ListTile(
      contentPadding: EdgeInsets.symmetric(horizontal: 16, vertical: 4),
      leading: GestureDetector(
        onTap: () => _navigateToProfile(user.username),
        child: CircleAvatar(
          radius: 24,
          backgroundImage: NetworkImage(user.avatar),
        ),
      ),
      title: GestureDetector(
        onTap: () => _navigateToProfile(user.username),
        child: Row(
          children: [
            Text(
              user.username,
              style: TextStyle(fontWeight: FontWeight.bold),
            ),
            if (user.isFollowedBy)
              Container(
                margin: EdgeInsets.only(left: 8),
                padding: EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                decoration: BoxDecoration(
                  color: Colors.grey[200],
                  borderRadius: BorderRadius.circular(4),
                ),
                child: Text(
                  'Follows you',
                  style: TextStyle(fontSize: 10, color: Colors.grey[600]),
                ),
              ),
          ],
        ),
      ),
      subtitle: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(user.displayName),
          Text(
            'Followed ${user.followedDate}',
            style: TextStyle(fontSize: 12, color: Colors.grey),
          ),
        ],
      ),
      trailing: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          IconButton(
            icon: Icon(Icons.message, size: 20),
            onPressed: () => _sendMessage(user),
          ),
          SizedBox(width: 8),
          _buildUnfollowButton(user),
        ],
      ),
    );
  }

  Widget _buildUnfollowButton(User user) {
    return SizedBox(
      width: 90,
      height: 32,
      child: OutlinedButton(
        onPressed: () => _showUnfollowDialog(user),
        style: OutlinedButton.styleFrom(
          side: BorderSide(color: Colors.grey),
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
          ),
        ),
        child: Text(
          'Following',
          style: TextStyle(
            color: Colors.black,
            fontSize: 12,
          ),
        ),
      ),
    );
  }

  void _showUnfollowDialog(User user) {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            CircleAvatar(
              radius: 40,
              backgroundImage: NetworkImage(user.avatar),
            ),
            SizedBox(height: 16),
            Text(
              'Unfollow ${user.username}?',
              style: TextStyle(
                fontSize: 18,
                fontWeight: FontWeight.bold,
              ),
            ),
            SizedBox(height: 8),
            Text(
              'You can always follow them again later.',
              style: TextStyle(color: Colors.grey),
              textAlign: TextAlign.center,
            ),
          ],
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              _unfollowUser(user);
            },
            child: Text(
              'Unfollow',
              style: TextStyle(
                color: Colors.red,
                fontWeight: FontWeight.bold,
              ),
            ),
          ),
        ],
      ),
    );
  }

  void _unfollowUser(User user) {
    setState(() {
      following.remove(user);
      filteredFollowing.remove(user);
    });

    // Show undo option
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('Unfollowed ${user.username}'),
        action: SnackBarAction(
          label: 'Undo',
          onPressed: () {
            setState(() {
              following.add(user);
              _filterFollowing();
            });
          },
        ),
        duration: Duration(seconds: 3),
      ),
    );
  }

  void _sortFollowing(String sortType) {
    setState(() {
      switch (sortType) {
        case 'alphabetical':
          following.sort((a, b) => a.username.compareTo(b.username));
          break;
        case 'recent':
          // In real app, sort by actual follow date
          following.sort((a, b) => a.followedDate.compareTo(b.followedDate));
          break;
        case 'earliest':
          following.sort((a, b) => b.followedDate.compareTo(a.followedDate));
          break;
      }
      _filterFollowing();
    });
  }

  Future<void> _refreshFollowing() async {
    setState(() {
      isLoading = true;
    });

    // Simulate API call
    await Future.delayed(Duration(seconds: 2));

    // In real app, fetch updated following list from server
    setState(() {
      isLoading = false;
    });
  }

  void _sendMessage(User user) {
    Navigator.pushNamed(
      context,
      '/chat',
      arguments: user.username,
    );
  }

  void _navigateToProfile(String username) {
    Navigator.pushNamed(
      context,
      '/user-profile',
      arguments: username,
    );
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }
}

class User {
  final String id;
  final String username;
  final String displayName;
  final String avatar;
  bool isFollowing;
  final bool isFollowedBy;
  final String followedDate;

  User({
    required this.id,
    required this.username,
    required this.displayName,
    required this.avatar,
    required this.isFollowing,
    required this.isFollowedBy,
    required this.followedDate,
  });
}

File: snippets/screens/swiftui/following_screen/2/following_screen.swift
import SwiftUI

struct FollowingScreen: View {
    let username: String
    @StateObject private var viewModel = FollowingViewModel()
    @State private var searchText = ""
    @State private var showingSortOptions = false

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search bar
                SearchBar(text: $searchText)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 12)

                // Category filters
                CategoryFilterBar(
                    selectedCategory: $viewModel.selectedCategory,
                    categories: viewModel.categories
                )

                // Following list
                List {
                    ForEach(viewModel.filteredFollowing) { user in
                        FollowingRow(
                            user: user,
                            onUnfollow: {
                                viewModel.showUnfollowConfirmation(for: user)
                            },
                            onMessage: {
                                // Navigate to chat
                            },
                            onUserTap: {
                                // Navigate to user profile
                            }
                        )
                        .listRowInsets(EdgeInsets(top: 8, leading: 16, bottom: 8, trailing: 16))
                    }
                }
                .listStyle(.plain)
                .refreshable {
                    await viewModel.refreshFollowing()
                }
            }
            .navigationTitle(viewModel.navigationTitle)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(action: { showingSortOptions = true }) {
                        Image(systemName: "arrow.up.arrow.down")
                    }
                }
            }
        }
        .onAppear {
            viewModel.loadFollowing(for: username)
        }
        .onChange(of: searchText) { text in
            viewModel.filterFollowing(searchText: text)
        }
        .onChange(of: viewModel.selectedCategory) { _ in
            viewModel.applyFilters()
        }
        .actionSheet(isPresented: $showingSortOptions) {
            ActionSheet(
                title: Text("Sort Following"),
                buttons: [
                    .default(Text("Sort A-Z")) { viewModel.sortFollowing(.alphabetical) },
                    .default(Text("Recently followed")) { viewModel.sortFollowing(.recent) },
                    .default(Text("Followed earliest")) { viewModel.sortFollowing(.earliest) },
                    .cancel()
                ]
            )
        }
        .alert("Unfollow User", isPresented: $viewModel.showUnfollowAlert) {
            Button("Cancel", role: .cancel) { }
            Button("Unfollow", role: .destructive) {
                viewModel.confirmUnfollow()
            }
        } message: {
            if let user = viewModel.userToUnfollow {
                Text("Are you sure you want to unfollow \(user.username)?")
            }
        }
    }
}

struct FollowingRow: View {
    let user: FollowingUserModel
    let onUnfollow: () -> Void
    let onMessage: () -> Void
    let onUserTap: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Button(action: onUserTap) {
                AsyncImage(url: URL(string: user.avatar)) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 48, height: 48)
                .clipShape(Circle())
            }

            VStack(alignment: .leading, spacing: 2) {
                HStack {
                    Button(action: onUserTap) {
                        Text(user.username)
                            .font(.headline)
                            .fontWeight(.semibold)
                            .foregroundColor(.primary)
                    }

                    if user.isFollowedBy {
                        Text("Follows you")
                            .font(.caption)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color.gray.opacity(0.2))
                            .cornerRadius(4)
                            .foregroundColor(.secondary)
                    }
                }

                Text(user.displayName)
                    .font(.subheadline)
                    .foregroundColor(.secondary)

                Text("Followed \(user.followedDate)")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            Spacer()

            HStack(spacing: 8) {
                Button(action: onMessage) {
                    Image(systemName: "message")
                        .font(.system(size: 16))
                        .foregroundColor(.blue)
                }

                Button("Following") {
                    onUnfollow()
                }
                .font(.subheadline)
                .foregroundColor(.primary)
                .padding(.horizontal, 16)
                .padding(.vertical, 6)
                .background(Color.gray.opacity(0.1))
                .cornerRadius(8)
                .overlay(
                    RoundedRectangle(cornerRadius: 8)
                        .stroke(Color.gray.opacity(0.3), lineWidth: 1)
                )
            }
        }
    }
}

@MainActor
class FollowingViewModel: ObservableObject {
    @Published var following: [FollowingUserModel] = []
    @Published var filteredFollowing: [FollowingUserModel] = []
    @Published var selectedCategory = "All"
    @Published var navigationTitle = "Following"
    @Published var showUnfollowAlert = false
    @Published var userToUnfollow: FollowingUserModel?

    let categories = ["All", "Following you back", "Not following back", "Recently followed"]

    private var allFollowing: [FollowingUserModel] = []
    private var currentSearchText = ""

    enum SortType {
        case alphabetical, recent, earliest
    }

    func loadFollowing(for username: String) {
        // Sample following data
        allFollowing = [
            FollowingUserModel(
                id: "1",
                username: "alice_wonder",
                displayName: "Alice Wonder",
                avatar: "https://example.com/avatar3.jpg",
                isFollowing: true,
                isFollowedBy: true,
                followedDate: "2 months ago"
            ),
            FollowingUserModel(
                id: "2",
                username: "bob_builder",
                displayName: "Bob Builder",
                avatar: "https://example.com/avatar4.jpg",
                isFollowing: true,
                isFollowedBy: false,
                followedDate: "1 month ago"
            ),
            FollowingUserModel(
                id: "3",
                username: "charlie_brown",
                displayName: "Charlie Brown",
                avatar: "https://example.com/avatar5.jpg",
                isFollowing: true,
                isFollowedBy: true,
                followedDate: "3 weeks ago"
            ),
            FollowingUserModel(
                id: "4",
                username: "diana_prince",
                displayName: "Diana Prince",
                avatar: "https://example.com/avatar6.jpg",
                isFollowing: true,
                isFollowedBy: false,
                followedDate: "2 weeks ago"
            ),
            FollowingUserModel(
                id: "5",
                username: "edward_stark",
                displayName: "Edward Stark",
                avatar: "https://example.com/avatar7.jpg",
                isFollowing: true,
                isFollowedBy: true,
                followedDate: "1 week ago"
            ),
            FollowingUserModel(
                id: "6",
                username: "frank_castle",
                displayName: "Frank Castle",
                avatar: "https://example.com/avatar8.jpg",
                isFollowing: true,
                isFollowedBy: false,
                followedDate: "3 days ago"
            )
        ]

        following = allFollowing
        navigationTitle = "\(following.count) following"
        applyFilters()
    }

    func filterFollowing(searchText: String) {
        currentSearchText = searchText
        applyFilters()
    }

    func applyFilters() {
        var filtered = allFollowing

        // Apply category filter
        switch selectedCategory {
        case "Following you back":
            filtered = filtered.filter { $0.isFollowedBy }
        case "Not following back":
            filtered = filtered.filter { !$0.isFollowedBy }
        case "Recently followed":
            // Sort by most recent and take top 3
            filtered = filtered.sorted { user1, user2 in
                // In real app, use actual dates
                return user1.followedDate < user2.followedDate
            }.prefix(3).map { $0 }
        default:
            break
        }

        // Apply search filter
        if !currentSearchText.isEmpty {
            filtered = filtered.filter { user in
                user.username.localizedCaseInsensitiveContains(currentSearchText) ||
                user.displayName.localizedCaseInsensitiveContains(currentSearchText)
            }
        }

        filteredFollowing = filtered
    }

    func sortFollowing(_ sortType: SortType) {
        switch sortType {
        case .alphabetical:
            allFollowing.sort { $0.username < $1.username }
        case .recent:
            allFollowing.sort { $0.followedDate < $1.followedDate }
        case .earliest:
            allFollowing.sort { $0.followedDate > $1.followedDate }
        }
        applyFilters()
    }

    func showUnfollowConfirmation(for user: FollowingUserModel) {
        userToUnfollow = user
        showUnfollowAlert = true
    }

    func confirmUnfollow() {
        guard let user = userToUnfollow else { return }

        if let index = allFollowing.firstIndex(where: { $0.id == user.id }) {
            allFollowing.remove(at: index)
            following = allFollowing
            navigationTitle = "\(following.count) following"
            applyFilters()
        }

        userToUnfollow = nil
    }

    func refreshFollowing() async {
        // Simulate network request
        try? await Task.sleep(nanoseconds: 1_000_000_000)
        // Refresh data if needed
    }
}

struct FollowingUserModel: Identifiable {
    let id: String
    let username: String
    let displayName: String
    let avatar: String
    var isFollowing: Bool
    let isFollowedBy: Bool
    let followedDate: String
}

#Preview {
    FollowingScreen(username: "john_doe")
}